src/mqtthandler.rs

```diff
@@ -105,20 +151,69 @@ pub fn handle_subscribe<'a>(ctx: &mut Context, p: Subscribe<'a>) -> Vec<Packet<'
         }
     }
 
-    let reason = if allowed {
-        ctx.state.subs.insert(p.topic.to_owned());
+    if !allowed {
+        return vec![Packet::SubAck(SubAck {
+            id: p.id,
+            reason: Reason::NotAuthorized,
+        })];
+    }
+
+    let mut retained = None;
+
+    match ctx.storage.read_retained(p.topic, None) {
+        Ok(Some(r)) => retained = Some(r),
+        Ok(None) | Err(StorageError::StoreNotFound) => {}
+        Err(_) => {
+            return vec![Packet::SubAck(SubAck {
+                id: p.id,
+                reason: Reason::UnspecifiedError,
+            })]
+        }
+    }
+
+    let version = retained.as_ref().map(|r| Version {
+        generation: r.version.generation,
+        seq: r.version.seq,
+    });
+
+    ctx.state.subs.insert(
+        p.topic.to_string(),
+        Subscription {
+            no_local: p.no_local,
+            retain_as_published: p.retain_as_published,
+            last: Some(Last { version }),
+            ignore: Vec::new(),
+        },
+    );
 
-        Reason::Success
-    } else {
-        Reason::NotAuthorized
-    };
+    let mut out = Vec::new();
 
-    vec![Packet::SubAck(SubAck { id: p.id, reason })]
+    out.push(Packet::SubAck(SubAck {
+        id: p.id,
+        reason: Reason::Success,
+    }));
```

Comment:
Another small nit, just curious why we don't initiate the vec with the first item?

```suggestion
    let mut out = vec![Packet::SubAck(SubAck {
        id: p.id,
        reason: Reason::Success,
    })];
```